src/main/rascal/refactor/Rename.rsc

```diff
@@ -29,200 +29,176 @@ module refactor::Rename
+RenameResult rename(
+        list[Tree] cursor
+      , str newName
+      , Tree(loc) parseLoc
+      , TModel(Tree) tmodelForTree
+      , set[Define](list[Tree] cursor, TModel(Tree) getTModel, Renamer r) findDefinitions
+      , set[loc](set[Define] defs, Renamer r) findCandidateFiles
+      , void(Define def, str newName, TModel tm, Renamer r) renameDef
+      , bool(set[Define] defs, Tree t, Renamer r) skipCandidate = bool(_, _, _) { return false; }
+      , bool debug = true) {
+
+    // Tree & TModel caching
+
+    @memo{maximumSize(50)}
+    TModel getTModelCached(Tree t) = tmodelForTree(t);
+
+    @memo{maximumSize(50)}
+    Tree parseLocCached(loc l) {
+        // We already have the parse tree of the module under cursor
+        if (l == cursor[-1].src.top) {
+            return cursor[-1];
         }
-    };
 
-    list[ModelTask] modelTaskQueue = [];
-    list[tuple[loc, RenameState]] modelTasksDone = [];
-    solver.collectTModel = void(loc l, void(RenameState, TModel, RenameSolver) doWork, RenameState state) {
-        if (<l, state> notin modelTasksDone) {
-            modelTaskQueue += <l, doWork, state>;
-            modelTasksDone += <l, state>;
-        } else if (config.reportCollectCycles) {
-            println("-- Cycle detected: skipping TModel collection for <describeState(state)> (<l>)");
-        }
-    };
+        return parseLoc(l);
+    }
 
-    // REGISTER
+    // Messages
     set[Message] messages = {};
-    solver.msg = void(Message msg) {
+    void registerMessage(Message msg) {
         messages += msg;
     };
 
-    lrel[loc file, DocumentEdit edit] docEdits = [];
-    solver.documentEdit = void(DocumentEdit edit) {
-        loc f = edit has file ? edit.file : edit.from;
-        docEdits += <f, edit>;
-    };
-
-    solver.textEdit = void(TextEdit edit) {
-        loc f = edit.range.top;
-        docEdits += <f, changed(f, [edit])>;
-    };
-
-    map[str id, ChangeAnnotation annotation] annotations = ();
-    solver.annotation = void(str annotationId, ChangeAnnotation annotation) {
-        if (annotationId in annotations) throw "An annotation with id \'<annotationId>\' already exists!";
-        annotations[annotationId] = annotation;
-    };
-
-    // STORE
-    map[str, value] store = ();
-    solver.readStore = value(str key) { return store[key]; };
-    solver.writeStore = void(str key, value val) {
-        store[key] = val;
-    };
-
-    // RUN
-    map[loc, Tree] treeCache = ();
-    map[loc, TModel] modelCache = ();
+    // Edits
+    set[value] editsSeen = {};
+    list[DocumentEdit] docEdits = [];
 
-    Tree getTree(loc l) {
-        if (l notin treeCache) {
-            treeCache[l] = config.parseLoc(l);
-        } else if (config.debug) {
-            println("-- Using cached tree for <l>");
+    void checkEdit(te:replace(loc range, _)) {
+        if (te in editsSeen) {
+            messages += error("Multiple replace edits for this location.", range);
```

Comment:
what if the `loc` is the same, but the replacement isn't? maybe the `editSeen` should be a `set[loc]`?